CONTRIBUTING.rst

```diff
@@ -1,47 +1,179 @@
+This project, hosted on https://github.com/nilearn/nilearn, is a community
+effort, and everyone is welcome to contribute. We value very much your feedback
+and opinion on features that should be improved or added. All discussions
+are public and hold on relevant issues or pull requests. To discuss your matter,
+please answer a relevant `issue <https://github.com/nilearn/nilearn/issues>`_
+or open a new one.
 
 The best way to contribute and to help the project is to start working on known
-issues.
-See `Good first issues <https://github.com/nilearn/nilearn/labels/Good%20first%20issue>`_ to get
-started.
+issues such as `Good first issues <https://github.com/nilearn/nilearn/labels/Good%20first%20issue>`_ ,
+`known bugs <https://github.com/nilearn/nilearn/labels/Bug>`_ or
+`proposed enhancements <https://github.com/nilearn/nilearn/labels/Enhancement>`_.
 
 If an issue does not already exist for a potential contribution, we ask that
 you first open an `issue <https://github.com/nilearn/nilearn/issues>`_ before
-sending a :ref:`pull request`.
+sending a :ref:`pull request` to discuss in advance scope and potential design
+choices.
 
-Opening an issue
-------------------
 
-Nilearn uses issues for tracking bugs, requesting potential features, and
-holding project discussions.
+Contribution Guidelines
+---------------------------
 
-Core developers can assign labels on issues, such as:
+We ask every contributor to respect common guidelines. Those are inspired from
+`scikit-learn
+<https://scikit-learn.org/stable/developers/contributing.html#contributing-code>`_
+and ensure Nilearn remains simple to understand, efficient and maintainable.
+For example tests need to run quickly in order not to burden the development process.
+To keep continuous integration efficient with our limited infrastructure, running
+all the examples must lead to downloading a limited amount of data (gigabytes)
+and execute in a reasonable amount of time (a few hours).
+
+Those guidelines will hence be checked during reviewing process.
+
+
++--------------+-------------+---------------------------------------+
+|              | Which PR ?  |        Guidelines                     |
++==============+=============+=======================================+
+|              |             | - Clearly showcase benefits           |
+|  Examples    | New features| - Run in a few minutes                |
+|              |             | - Use light data from Nilearn         |
+|              |             | - Renders well after build            |
++--------------+-------------+---------------------------------------+
+|              |             | - Test type is adapted to behavior    |
+|              |             | - Tests pass continuous integration   |
+|              |  Bugfixes   | - Doesn't decrease coverage           |
+|    Tests     | New features| - Fast, using small mocked data       |
+|              |             | - Atomic (one per function) and seeded|
```

Comment:
maybe "and random number generators should be seeded" ? adding more keywords will make it easier to look up for someone who doesn't know what this means

Reply:
I'm going to include that below I think with the precise command. I remember you standardized them recently, what command should be used ?

Reply:
do you mean #2741?
basically,

if you need to explicitly generate random numbers, use a seeded generator:
good:
```
rng = np.random.RandomState(0)
my_number = rng.normal()
```

bad:
```
my_number = np.random.normal()
```
also bad:
```
np.random.seed(0)
my_number = np.random.normal()
```

note that in recent(ish) numpy versions the correct way is
```
rng = np.random.default_rng(0)
my_number = rng.normal()
```

but we still support old numpy versions that don't have `default_rng`.

and if you use some object that asks for some random seed, eg a scikit-learn estimator, pass either a `np.random.RandomState` or an int as the seed, not the default `None`

good

```
kmeans = sklearn.cluster.KMeans(random_state=0)
```
or
```
rng = np.random.RandomState(0)
kmeans = sklearn.cluster.KMeans(random_state=rng)
```

bad

```
kmeans = sklearn.cluster.KMeans()
```

and of course the seeded generator will produce the same sequence every time (that's the whole point), so if you need to apply a function to a few different random sequences, either pass a different seed or reuse the same generator.

```
x1 = make_some_fake_data(random_state=0)
x2 = make_some_fake_data(random_state=0) # x1 == x2
```
so
```
rng = np.random.RandomState(0)
x1 = make_some_fake_data(random_state=rng)
x2 = make_some_fake_data(random_state=rng)
```

or 
```
x1 = make_some_fake_data(random_state=0)
x2 = make_some_fake_data(random_state=1)
```

Reply:
Thank you very much for the recap. I'll add those good practices (as of today) in the doc as I try to put all those conventions in one place !